presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -15,283 +15,367 @@
+            FlatMapFunction2<Iterator<Tuple2<Integer, PrestoSparkRow>>, Iterator<Tuple2<Integer, PrestoSparkRow>>, Tuple2<Integer, PrestoSparkRow>> taskProcessor =
+                    createTaskProcessor(
+                            executorFactoryProvider,
+                            serializedTaskDescriptor,
+                            firstRemoteSource.getId().toString(),
+                            secondRemoteSource.getId().toString(),
+                            taskStatsCollector);
+            return JavaPairRDD.fromJavaRDD(
+                    firstRdd.zipPartitions(
+                            secondRdd,
+                            taskProcessor));
         }
 
-        private static List<List<ScheduledSplit>> assignSplitsToTasks(List<ScheduledSplit> scheduledSplits, int numTasks)
-        {
-            List<List<ScheduledSplit>> assignedSplits = new ArrayList<>();
-            for (int i = 0; i < numTasks; i++) {
-                assignedSplits.add(new ArrayList<>());
-            }
+        throw new IllegalArgumentException(format("unsupported number of inputs: %s", rddInputs.size()));
+    }
 
-            for (ScheduledSplit split : scheduledSplits) {
-                int taskId = Objects.hash(split.getPlanNodeId(), split.getSequenceId()) % numTasks;
-                if (taskId < 0) {
-                    taskId += numTasks;
-                }
+    private JavaPairRDD<Integer, PrestoSparkRow> createSourceRdd(
+            JavaSparkContext sparkContext,
+            Session session,
+            PlanFragment fragment,
+            PrestoSparkTaskExecutorFactoryProvider executorFactoryProvider,
+            CollectionAccumulator<SerializedTaskStats> taskStatsCollector,
+            TableWriteInfo tableWriteInfo)
+    {
+        // TODO: Possible in case of a broadcast join
+        checkArgument(fragment.getRemoteSourceNodes().isEmpty(), "source task with remote sources is not supported");
+
+        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
+        checkArgument(
+                tableScans.size() == 1,
+                "exactly one table scan is expected in SOURCE_DISTRIBUTION fragment. fragmentId: %s, actual number of table scans: %s",
+                fragment.getId(),
+                tableScans.size());
+        verify(tableScans.size() == fragment.getTableScanSchedulingOrder().size());
+
+        TableScanNode tableScan = tableScans.get(0);
+
+        List<ScheduledSplit> splits = getSplits(session, tableScan);
+        shuffle(splits);
+        int initialPartitionCount = getSparkInitialPartitionCount(session);
+        int numTasks = Math.min(splits.size(), initialPartitionCount);
+        if (numTasks == 0) {
+            return JavaPairRDD.fromJavaRDD(sparkContext.emptyRDD());
+        }
 
-                assignedSplits.get(taskId).add(split);
-            }
+        List<List<ScheduledSplit>> assignedSplits = assignSplitsToTasks(splits, numTasks);
```

Comment:
nit: Add a TODO state we want to refactor `List<List<ScheduledSplit>>` to be a dedicated class such as `SplitAssignment`. 

The motivation is while I suggest this part of code, when I reread my own code my first question is : hmm... what's this List of List? ...

Reply:
Frankly It feels like this code is pretty local. I'm not sure it really deserves a separate class. Let's keep it as is for now.